cpp/include/raft/neighbors/detail/refine_host-inl.hpp

```diff
@@ -43,35 +43,43 @@ template <typename DC, typename IdxT, typename DataT, typename DistanceT, typena
     "neighbors::refine_host(%zu, %zu -> %zu)", n_queries, orig_k, refined_k);
 
   auto suggested_n_threads = std::max(1, std::min(omp_get_num_procs(), omp_get_max_threads()));
-  if (size_t(suggested_n_threads) > n_queries) { suggested_n_threads = n_queries; }
 
-#pragma omp parallel num_threads(suggested_n_threads)
-  {
-    std::vector<std::tuple<DistanceT, IdxT>> refined_pairs(orig_k);
-    for (size_t i = omp_get_thread_num(); i < n_queries; i += omp_get_num_threads()) {
-      // Compute the refined distance using original dataset vectors
+  // For efficiency, each thread should read a certain amount of array elements.
+  // The number of threads for distance computation is determined taking this into account.
+  constexpr int n_elements = 512;
+  size_t max_n_threads = ((n_queries * orig_k * dim) + n_elements - 1) / n_elements;
+  auto suggested_n_threads_for_distance = std::min(size_t(suggested_n_threads), max_n_threads);
+
+  // The max number of threads for topk computation is the number of queries.
+  auto suggested_n_threads_for_topk = std::min(size_t(suggested_n_threads), n_queries);
+
+  std::vector<std::vector<std::tuple<DistanceT, IdxT>>>
+      refined_pairs(n_queries, std::vector<std::tuple<DistanceT, IdxT>>(orig_k));
```

Comment:
With the PR we allocate a whole n_queries*orig_k array here rather than keeping one independent orig_k array per thread. Have you checked if this has any adverse effect on performance with large batches (e.g. n_queries 10K)?

Reply:
Yes, the larger the batch size, the larger the size of `refined_pairs`, so performance may be degraded if the batch size is very large. I'll check and add a workaround (blocking) if there is a problem.

Reply:
As per your concern, I found that using this implementation when the batch size is large results in lower performance, so I have changed it to use this finer-grained thread parallelism only when the number of queries is small.